flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/SharedBuffer.java

```diff
@@ -256,16 +267,75 @@ public void prune(long pruningTimestamp) {
 										edge.getTarget(),
 										edge.getVersion(),
 										copy));
+								if (matchingBehaviour == MatchingBehaviour.AFTER_LAST) {
+									cleanUp.add(edge.getTarget());
+								}
 							}
 						}
 					}
 				}
 			}
 		}
 
+		// Remove shared buffer entries to maintain correct matching behaviour
+		doCleanUp(new Predicate<K, V>() {
+
+			@Override
+			public boolean toRemove(SharedBufferEntry<K, V> entry) {
+				return cleanUp.contains(entry);
+			}
+		});
+		// Remove all entries that are dependent on the current event
+		if (matchingBehaviour == MatchingBehaviour.AFTER_LAST) {
+			doCleanUp(new Predicate<K, V>() {
+
+				@Override
+				public boolean toRemove(SharedBufferEntry<K, V> entry) {
+					if (entry == null) {
+						return false;
+					}
+					return entry.getValueTime().value == value
+						&& entry.getValueTime().timestamp == timestamp;
+				}
+			});
+		}
+
 		return result;
 	}
 
+	private void doCleanUp(Predicate<K, V> predicate) {
+		ArrayList<SharedBufferEntry<K, V>> toRemove = new ArrayList<>();
+		for (SharedBufferPage<K, V> page : this.pages.values()) {
+			for (SharedBufferEntry<K, V> entry : page.getEntries()) {
+				if (entry.getReferenceCounter() <= 1) {
+					doRecursiveCleanup(entry, predicate, toRemove);
+				}
+			}
+		}
+
+		for (SharedBufferEntry<K, V> startNode: toRemove) {
+			release(startNode.page.getKey(), startNode.getValueTime().value, startNode.getValueTime().getTimestamp());
+			remove(startNode.page.getKey(), startNode.getValueTime().value, startNode.getValueTime().getTimestamp());
+		}
+	}
+
+	private boolean doRecursiveCleanup(SharedBufferEntry<K, V> startNode, Predicate<K, V> cleanUp, ArrayList<SharedBufferEntry<K, V>> toRemove) {
```

Comment:
Replace ArrayList by List in the arguments, unless we need it to be ArrayList explicitly.